packages/lib/services/survey.ts

```diff
@@ -78,183 +82,245 @@ export const preloadSurveyWithAnalytics = (surveyId: string) => {
-    }
+      if (!surveyPrisma) {
+        throw new ResourceNotFoundError("Survey", surveyId);
+      }
 
-    if (!surveyPrisma) {
-      throw new ResourceNotFoundError("Survey", surveyId);
-    }
+      let { _count, displays, ...surveyPrismaFields } = surveyPrisma;
 
-    let { _count, displays, ...surveyPrismaFields } = surveyPrisma;
-
-    const numDisplays = displays.length;
-    const numDisplaysResponded = displays.filter((item) => item.status === "responded").length;
-    const numResponses = _count.responses;
-    // responseRate, rounded to 2 decimal places
-    const responseRate = numDisplays ? Math.round((numDisplaysResponded / numDisplays) * 100) / 100 : 0;
-
-    const transformedSurvey = {
-      ...surveyPrismaFields,
-      triggers: surveyPrismaFields.triggers.map((trigger) => trigger.eventClass),
-      analytics: {
-        numDisplays,
-        responseRate,
-        numResponses,
-      },
-    };
+      const numDisplays = displays.length;
+      const numDisplaysResponded = displays.filter((item) => item.status === "responded").length;
+      const numResponses = _count.responses;
+      // responseRate, rounded to 2 decimal places
+      const responseRate = numDisplays ? Math.round((numDisplaysResponded / numDisplays) * 100) / 100 : 0;
 
-    try {
-      const survey = ZSurveyWithAnalytics.parse(transformedSurvey);
-      return survey;
-    } catch (error) {
-      console.log(error);
-      if (error instanceof z.ZodError) {
-        console.error(JSON.stringify(error.errors, null, 2)); // log the detailed error information
-      }
-      throw new ValidationError("Data validation of survey failed");
-    }
-  }
-);
+      const transformedSurvey = {
+        ...surveyPrismaFields,
+        triggers: surveyPrismaFields.triggers.map((trigger) => trigger.eventClass),
+        analytics: {
+          numDisplays,
+          responseRate,
+          numResponses,
+        },
+      };
 
-export const getSurvey = cache(async (surveyId: string): Promise<TSurvey | null> => {
-  validateInputs([surveyId, ZId]);
-  let surveyPrisma;
-  try {
-    surveyPrisma = await prisma.survey.findUnique({
-      where: {
-        id: surveyId,
-      },
-      select: selectSurvey,
-    });
-  } catch (error) {
-    if (error instanceof Prisma.PrismaClientKnownRequestError) {
-      throw new DatabaseError("Database operation failed");
+      try {
+        const survey = ZSurveyWithAnalytics.parse(transformedSurvey);
+        return survey;
+      } catch (error) {
+        console.log(error);
+        if (error instanceof z.ZodError) {
+          console.error(JSON.stringify(error.errors, null, 2)); // log the detailed error information
+        }
+        throw new ValidationError("Data validation of survey failed");
+      }
+    },
+    [getSurveyWithAnalyticsCacheTag(surveyId)],
+    {
+      tags: [getSurveyWithAnalyticsCacheTag(surveyId)],
     }
+  )();
 
-    throw error;
-  }
-
-  if (!surveyPrisma) {
+  if (!survey) {
```

Comment:
Can survey be something else then `null`? Because if not then we can remove this block? ;-)

Reply:
same for other cached functions